pkg/logs/metrics/capacity_monitor.go

```diff
@@ -0,0 +1,81 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-present Datadog, Inc.
+
+package metrics
+
+import (
+	"sync"
+	"time"
+)
+
+// CapacityMonitor samples the average capacity of a component over a given interval.
+// Capacity is calculated as the difference between the ingress and egress of a payload.
+// Because data moves very quickly through components, we need to sample and aggregate this value over time.
+type CapacityMonitor struct {
+	sync.Mutex
+	ingress      int64
+	ingressBytes int64
+	egress       int64
+	egressBytes  int64
+	avgItems     float64
+	avgBytes     float64
+	name         string
+	instance     string
+	tickChan     <-chan time.Time
+}
+
+// NewCapacityMonitor creates a new CapacityMonitor
+func NewCapacityMonitor(name, instance string) *CapacityMonitor {
+	return newCapacityMonitorWithTick(name, instance, time.NewTicker(1*time.Second).C)
+}
+
+// newCapacityMonitorWithTick is used for testing.
+func newCapacityMonitorWithTick(name, instance string, tickChan <-chan time.Time) *CapacityMonitor {
+	return &CapacityMonitor{
+		name:     name,
+		instance: instance,
+		avgItems: 0,
+		avgBytes: 0,
+		tickChan: tickChan,
+	}
+}
+
+// AddIngress records the ingress of a payload
+func (i *CapacityMonitor) AddIngress(pl MeasurablePayload) {
+	i.Lock()
+	defer i.Unlock()
+	i.ingress += pl.Count()
+	i.ingressBytes += pl.Size()
+	i.sample()
+}
+
+// AddEgress records the egress of a payload
+func (i *CapacityMonitor) AddEgress(pl MeasurablePayload) {
+	i.Lock()
+	defer i.Unlock()
+	i.egress += pl.Count()
+	i.egressBytes += pl.Size()
+	i.sample()
+
+}
+
+func (i *CapacityMonitor) sample() {
+	select {
+	case <-i.tickChan:
+		i.avgItems = ewma(float64(i.ingress-i.egress), i.avgItems)
```

Comment:
We discussed this elsewhere already, but just for the record: a Histogram is probably more appropriate choice for a fast moving metric like this, since it will also capture short spikes that can be missed by periodic sampling like this, that would still nonetheless contribute to the agent's memory usage.